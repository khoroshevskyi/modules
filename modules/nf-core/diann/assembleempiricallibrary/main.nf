process DIANN_ASSEMBLEEMPIRICALLIBRARY {
    tag "$meta.id"
    label 'process_low'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://containers.biocontainers.pro/s3/SingImgsRepo/diann/v1.8.1_cv1/diann_v1.8.1_cv1.img' :
        'docker.io/biocontainers/diann:v1.8.1_cv1' }"

    input:
    tuple val(meta), path(ms_files), path(lib), path("quant/*")

    output:
    tuple val(meta), path("empirical_library.*"), emit: empirical_library
    tuple val(meta), path("assemble_empirical_library.log"), emit: log
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    // Exit if running this module with -profile conda / -profile mamba
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        error "DIANN_ASSEMBLEEMPIRICALLIBRARY module does not support Conda. Please use Docker / Singularity / Podman instead."
    }
    def args = task.ext.args ?: ''

    """
    ls -lcth

    diann   --f ${(ms_files as List).join(' --f ')} \\
            --lib ${lib} \\
            --threads ${task.cpus} \\
            --out-lib empirical_library \\
            --temp ./quant/ \\
            ${args}

    cp report.log.txt assemble_empirical_library.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        DIA-NN: \$(diann 2>&1 | grep "DIA-NN" | grep -oP "\\d+\\.\\d+(\\.\\w+)*(\\.[\\d]+)?")
    END_VERSIONS
    """

    stub:
    // Exit if running this module with -profile conda / -profile mamba
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        error "DIANN_ASSEMBLEEMPIRICALLIBRARY module does not support Conda. Please use Docker / Singularity / Podman instead."
    }
    def args = task.ext.args ?: ''

    """
    touch empirical_library.tsv
    touch assemble_empirical_library.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        DIA-NN: \$(diann 2>&1 | grep "DIA-NN" | grep -oP "\\d+\\.\\d+(\\.\\w+)*(\\.[\\d]+)?")
    END_VERSIONS
    """
}
