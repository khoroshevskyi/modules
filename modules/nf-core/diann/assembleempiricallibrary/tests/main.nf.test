nextflow_process {

    name "Test Process DIANN_ASSEMBLEEMPIRICALLIBRARY"
    script "../main.nf"
    process "DIANN_ASSEMBLEEMPIRICALLIBRARY"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "diann"
    tag "diann/assembleempiricallibrary"
    tag "untar"

    setup {
        run("UNTAR") {
            script "../../../../nf-core/untar/main.nf"
            process {
                """
                input[0] = [
                    [id:'test_exp'],
                    file(params.modules_testdata_base_path + 'proteomics/diann/RD139_Narrow_UPS1_0_1fmol_inj1.mzML.tar.gz', checkIfExists: true)
                ]
                """
            }
        }
    }

    test("proteomics - assemble empirical library") {

        when {
            process {
                """
                input[0] = UNTAR.out.untar.map{ meta, dir ->  
                    def mzmlFile = dir.resolve('RD139_Narrow_UPS1_0_1fmol_inj1.mzML')
                    [[experiment_id:'test_exp'], [mzmlFile], [file(params.modules_testdata_base_path + 'proteomics/diann/preliminary_analysis.quant', checkIfExists: true)], file(params.modules_testdata_base_path + 'proteomics/diann/lib.predicted.speclib', checkIfExists: true)]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert path("${process.out.log[0][1]}").exists() },
                { assert snapshot(
                    process.out.empirical_library,
                    process.out.versions
                ).match() }
            )
        }
    }

    test("proteomics - assemble empirical library - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = UNTAR.out.untar.map{ meta, dir ->  
                    def mzmlFile = dir.resolve('RD139_Narrow_UPS1_0_1fmol_inj1.mzML')
                    [[experiment_id:'test_exp'], [mzmlFile], [file(params.modules_testdata_base_path + 'proteomics/diann/preliminary_analysis.quant', checkIfExists: true)], file(params.modules_testdata_base_path + 'proteomics/diann/lib.predicted.speclib', checkIfExists: true)]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
