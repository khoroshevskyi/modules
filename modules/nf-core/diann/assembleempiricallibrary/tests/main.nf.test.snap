{
    "proteomics - assemble empirical library": {
        "content": [
            [
                [
                    {
                        "experiment_id": "test_exp"
                    },
                    "empirical_library.speclib:md5,bb01028baf4fd5be5f71d0238e6b7902"
                ]
            ],
            [
                "versions.yml:md5,8df80648eef272291b5ad162efcc95d4"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-24T14:34:04.930953"
    },
    "proteomics - assemble empirical library - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "empirical_library.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "assemble_empirical_library.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,8df80648eef272291b5ad162efcc95d4"
                ],
                "empirical_library": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "empirical_library.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "log": [
                    [
                        {
                            "experiment_id": "test_exp"
                        },
                        "assemble_empirical_library.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8df80648eef272291b5ad162efcc95d4"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-24T14:34:09.954123"
    }
}